spv/rescan.go

```diff
@@ -0,0 +1,129 @@
+// Copyright (c) 2013-2014 The btcsuite developers
+// Copyright (c) 2018-2019 The Decred developers
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+
+package spv
+
+import (
+	"github.com/decred/dcrd/blockchain/stake/v2"
+	"github.com/decred/dcrd/gcs/blockcf"
+	"github.com/decred/dcrd/txscript/v2"
+	"github.com/decred/dcrd/wire"
+)
+
+// rescanCheckTransaction is a helper function to rescan both stake and regular
+// transactions in a block.  It appends transactions that match the filters to
+// *matches, while updating the filters to add outpoints for new UTXOs
+// controlled by this wallet.  New data added to the Syncer's filters is also
+// added to fadded.
+//
+// This function may only be called with the filter mutex held.
+func (s *Syncer) rescanCheckTransactions(matches *[]*wire.MsgTx, fadded *blockcf.Entries, txs []*wire.MsgTx, tree int8, walletID int) {
+	for i, tx := range txs {
+		// Keep track of whether the transaction has already been added
+		// to the result.  It shouldn't be added twice.
+		added := false
+
+		txty := stake.TxTypeRegular
+		if tree == wire.TxTreeStake {
+			txty = stake.DetermineTxType(tx)
+		}
+
+		// Coinbases and stakebases are handled specially: all inputs of a
+		// coinbase and the first (stakebase) input of a vote are skipped over
+		// as they generate coins and do not reference any previous outputs.
+		inputs := tx.TxIn
+		if i == 0 && txty == stake.TxTypeRegular {
+			goto LoopOutputs
+		}
+		if txty == stake.TxTypeSSGen {
+			inputs = inputs[1:]
+		}
+
+		for _, input := range inputs {
+			if !s.rescanFilter[walletID].ExistsUnspentOutPoint(&input.PreviousOutPoint) {
+				continue
+			}
+			if !added {
+				*matches = append(*matches, tx)
+				added = true
+			}
+		}
+
+	LoopOutputs:
+		for i, output := range tx.TxOut {
+			_, addrs, _, err := txscript.ExtractPkScriptAddrs(
+				output.Version, output.PkScript,
+				s.chainParams)
```

Comment:
Instead of having the chainparams saved as a property of the `Syncer` struct, why not read it from the current wallet using the walletID property passed into this method?
```suggestion
				s.wallets[walletID].ChainParams())
```